source/mesh_deformation/fastscape.cc

```diff
@@ -0,0 +1,1237 @@
+                  }
+                while (istep<steps);
+
+                // Output out how long FastScape took to run.
+                auto t_end = std::chrono::high_resolution_clock::now();
+                double r_time = std::chrono::duration<double>(t_end-t_start).count();
+                this->get_pcout() << "      FastScape runtime... " << round(r_time*1000)/1000 << "s" << std::endl;
+              }
+
+              visualization_step = current_timestep;
+              if (make_vtk)
+              {
+                 this->get_pcout() << "      Writing VTK..." << std::endl;
+                 fastscape_named_vtk_(h.get(), &vexp, &visualization_step, c, &length);
+              }
+
+              // If we've reached the end time, destroy fastscape.
+              if (this->get_time()+a_dt >= end_time)
+                {
+                  this->get_pcout() << "      Destroying FastScape..." << std::endl;
+                  fastscape_destroy_();
+                }
+
+              // Find out our velocities from the change in height.
+              // Where V is a vector of array size that exists on all processors.
+              for (int i=0; i<array_size; i++)
+                {
+                  V[i] = (h[i] - h_old[i])/a_dt;
+                }
+
+              MPI_Bcast(&V[0], array_size, MPI_DOUBLE, 0, this->get_mpi_communicator());
+            }
+          else
+            {
+              for (unsigned int i=0; i<temporary_variables.size(); i++)
+                MPI_Ssend(&temporary_variables[i][0], temporary_variables[1].size(), MPI_DOUBLE, 0, 42, this->get_mpi_communicator());
+
+              MPI_Bcast(&V[0], array_size, MPI_DOUBLE, 0, this->get_mpi_communicator());
+            }
+
+          // Get the sizes needed for the data table.
+          TableIndices<dim> size_idx;
+          for (unsigned int d=0; d<dim; ++d)
+            {
+              size_idx[d] = table_intervals[d]+1;
+            }
+
+          // Initialize a table to hold all velocity values that will be interpolated back to ASPECT.
+          Table<dim,double> data_table;
+          data_table.TableBase<dim,double>::reinit(size_idx);
+          TableIndices<dim> idx;
+
+          // Loop through the data table and fill out the surface (i or k = 1) to the velocities from FastScape.
+          if (dim == 2)
+            {
+              std::vector<double> V2(nx);
+
+              for (int i=1; i<(nx-1); i++)
+                {
+                  // If using the center slice, find velocities from the row closest to the center.
+                  if (slice)
+                    {
+                      const int index = i+nx*(round((ny-1)/2));
+                      V2[i-1] = V[index];
+                    }
+                  // Here we use average velocities across the y nodes, excluding the ghost nodes (top and bottom row).
+                  // Note: If ghost nodes are turned off, boundary effects may influence this.
```

Comment:
Does this require a AssertThrow to make sure users don't turn off the ghost nodes?

Reply:
Hmmm, I'm not sure. I think it is nice to have the option to turn them off. But I'm not sure we've ever really turned them off except for in cases where we were testing something within the coupling.